core/src/Kokkos_ScratchSpace.hpp

```diff
@@ -100,87 +98,41 @@ class ScratchMemorySpace {
   template <typename IntType>
   KOKKOS_INLINE_FUNCTION void* get_shmem(const IntType& size,
                                          int level = -1) const {
-    if (level == -1) level = m_default_level;
-    if (level == 0) {
-      void* tmp = m_iter_L0 + m_offset * align(size);
-      if (m_end_L0 < (m_iter_L0 += align(size) * m_multiplier)) {
-        m_iter_L0 -= align(size) * m_multiplier;  // put it back like it was
-#ifdef KOKKOS_ENABLE_DEBUG
-        // mfh 23 Jun 2015: printf call consumes 25 registers
-        // in a CUDA build, so only print in debug mode.  The
-        // function still returns nullptr if not enough memory.
-        printf(
-            "ScratchMemorySpace<...>::get_shmem: Failed to allocate "
-            "%ld byte(s); remaining capacity is %ld byte(s)\n",
-            long(size), long(m_end_L0 - m_iter_L0));
-#endif  // KOKKOS_ENABLE_DEBUG
-        tmp = nullptr;
-      }
-      return tmp;
-    } else {
-      void* tmp = m_iter_L1 + m_offset * align(size);
-      if (m_end_L1 < (m_iter_L1 += align(size) * m_multiplier)) {
-        m_iter_L1 -= align(size) * m_multiplier;  // put it back like it was
-#ifdef KOKKOS_ENABLE_DEBUG
-        // mfh 23 Jun 2015: printf call consumes 25 registers
-        // in a CUDA build, so only print in debug mode.  The
-        // function still returns nullptr if not enough memory.
-        printf(
-            "ScratchMemorySpace<...>::get_shmem: Failed to allocate "
-            "%ld byte(s); remaining capacity is %ld byte(s)\n",
-            long(size), long(m_end_L1 - m_iter_L1));
-#endif  // KOKKOS_ENABLE_DEBUG
-        tmp = nullptr;
-      }
-      return tmp;
-    }
+    return get_shmem_common</*aligned*/ false>(size, 1, level);
   }
 
-  KOKKOS_INLINE_FUNCTION
-  void* get_shmem_aligned(const ptrdiff_t size, const ptrdiff_t alignment,
-                          int level = -1) const {
+  template <typename IntType>
+  KOKKOS_INLINE_FUNCTION void* get_shmem_aligned(const IntType& size,
+                                                 const ptrdiff_t alignment,
+                                                 int level = -1) const {
+    return get_shmem_common</*aligned*/ true>(size, alignment, level);
+  }
+
+  template <bool aligned, typename IntType>
+  KOKKOS_INLINE_FUNCTION void* get_shmem_common(const IntType& size,
```

Comment:
Actually make that guy private or prefix with `impl_`

Reply:
Done.